libp2p/protocols/pubsub/gossipsub.nim

```diff
@@ -249,69 +244,75 @@ proc getGossipPeers(g: GossipSub): Table[string, ControlMessage] {.gcsafe.} =
       result[id].ihave.add(ihave)
 
 proc heartbeat(g: GossipSub) {.async.} =
-  while g.heartbeatRunning:
-    try:
-      trace "running heartbeat"
+  # don't lock on every iteration
+  withLock g.heartbeatLock:
+    while g.heartbeatRunning:
+      try:
+        trace "running heartbeat"
 
-      for t in toSeq(g.topics.keys):
-        await g.rebalanceMesh(t)
+        for t in toSeq(g.topics.keys):
+          await g.rebalanceMesh(t)
 
-      await g.dropFanoutPeers()
+        await g.dropFanoutPeers()
 
-      # replenish known topics to the fanout
-      for t in toSeq(g.fanout.keys):
-        g.replenishFanout(t)
+        # replenish known topics to the fanout
+        for t in toSeq(g.fanout.keys):
+          g.replenishFanout(t)
 
-      let peers = g.getGossipPeers()
-      var sent: seq[Future[void]]
-      for peer in peers.keys:
-        if peer in g.peers:
-          sent &= g.peers[peer].send(@[RPCMsg(control: some(peers[peer]))])
-      checkFutures(await allFinished(sent))
+        let peers = g.getGossipPeers()
+        var sent: seq[Future[void]]
+        for peer in peers.keys:
+          if peer in g.peers:
+            sent &= g.peers[peer].send(@[RPCMsg(control: some(peers[peer]))])
+        checkFutures(await allFinished(sent))
 
-      g.mcache.shift() # shift the cache
-    except CancelledError as exc:
-      raise exc
-    except CatchableError as exc:
-      trace "exception ocurred in gossipsub heartbeat", exc = exc.msg
+        g.mcache.shift() # shift the cache
+      except CancelledError as exc:
+        raise exc
+      except CatchableError as exc:
+        trace "exception ocurred in gossipsub heartbeat", exc = exc.msg
 
-    await sleepAsync(5.seconds)
+      await sleepAsync(GossipSubHeartbeatInterval)
 
 method handleDisconnect*(g: GossipSub, peer: PubSubPeer) =
   ## handle peer disconnects
-  procCall FloodSub(g).handleDisconnect(peer)
 
+  if isNil(peer):
+    return
+
+  let id = peer.id
   for t in toSeq(g.gossipsub.keys):
     if t in g.gossipsub:
-      g.gossipsub[t].excl(peer.id)
+      g.gossipsub[t].excl(id)
 
     libp2p_gossipsub_peers_per_topic_gossipsub
-      .set(g.gossipsub.getOrDefault(t).len.int64, labelValues = [t])
+      .set(g.gossipsub[t].len.int64, labelValues = [t])
 
   for t in toSeq(g.mesh.keys):
     if t in g.mesh:
-      g.mesh[t].excl(peer.id)
+      g.mesh[t].excl(id)
 
     libp2p_gossipsub_peers_per_topic_mesh
       .set(g.mesh[t].len.int64, labelValues = [t])
 
   for t in toSeq(g.fanout.keys):
     if t in g.fanout:
-      g.fanout[t].excl(peer.id)
+      g.fanout[t].excl(id)
 
     libp2p_gossipsub_peers_per_topic_fanout
       .set(g.fanout[t].len.int64, labelValues = [t])
 
+  procCall FloodSub(g).handleDisconnect(peer)
+
 method subscribePeer*(p: GossipSub,
```

Comment:
```
method subscribePeer*(
    p: GossipSub,  # 4 spaces here!
    conn: Connection) =...
```
is an indent style from python that works well - the 4 space indent sets the parameters apart from the body, but doesn't require adjustments when refactoring stuff like the name :)

Reply:
We should just use a formatter to enforce common style like this across the board.

Reply:
there exists no formatter, hence it makes sense to optimize the format so that it doesn't require readjustment and aligning